File: smartcore-wasi/python/lib/python/glue.py
from wasmer import engine, wasi, Store, Module, Instance
from wasmer_compiler_cranelift import Compiler
import os

def get_string_ptr(string, instance):
    prepared_string = bytes(string, 'utf-8')
    length_of_string = len(prepared_string) + 1
    # Allocate memory for the subject, and get a pointer to it.
    string_ptr = instance.exports.allocate(length_of_string)

    # Write the subject into the memory.
    memory = instance.exports.memory.uint8_view(string_ptr)
    memory[0:length_of_string] = prepared_string
    memory[length_of_string] = 0  # C-string terminates by NULL.
    return (string_ptr, length_of_string)

def get_string_from_ptr(ptr, instance):
    memory = instance.exports.memory.uint8_view(ptr)
    memory_length = len(memory)
    output = []
    nth = 0

    while nth < memory_length:
        byte = memory[nth]

        if byte == 0:
            break

        output.append(byte)
        nth += 1
    length_of_output = nth

    return (bytes(output).decode(), length_of_output)

relative_dir = 'lib/python'

wasm_bytes = open(f'{relative_dir}/smartcore_wasi_lib.wasm', 'rb').read()
store = Store(engine.JIT(Compiler))
module = Module(store, wasm_bytes)
wasi_version = wasi.get_version(module, strict=True)
wasi_env = wasi.StateBuilder(
    'smartcore-wasi-lib').preopen_directory(".").finalize()

import_object = wasi_env.generate_import_object(store, wasi_version)

instance = Instance(module, import_object)

def load_model():
    (file_ptr, file_len) = get_string_ptr(f'{relative_dir}/iris_knn.model', instance)
    output_ptr = instance.exports.load_model(file_ptr)
    (output, output_len) = get_string_from_ptr(output_ptr, instance)
    instance.exports.deallocate(output_ptr, output_len)
    instance.exports.deallocate(file_ptr, file_len)
    return output

File: smartcore-wasi/go/lib/go/glue.go
package glue

import (
	"io/ioutil"

	wasmer "github.com/wasmerio/wasmer-go/wasmer"
)

var Instance *wasmer.Instance

func getStringPointer(data string, instance *wasmer.Instance) (int32, int) {
	// Set the subject to greet.
	dataPrepared := []byte(data)
	lengthOfSubject := len(dataPrepared) + 1

	// Allocate memory for the subject, and get a pointer to it.
	allocate, _ := instance.Exports.GetFunction("allocate")
	allocateResult, _ := allocate(lengthOfSubject)
	ptr := allocateResult.(int32)

	// Write the subject into the memory.
	memory, _ := instance.Exports.GetMemory("memory")
	memoryData := memory.Data()[ptr:]

	for nth := 0; nth < lengthOfSubject-1; nth++ {
		memoryData[nth] = dataPrepared[nth]
	}

	// C-string terminates by NULL.
	memoryData[lengthOfSubject-1] = 0

	return ptr, lengthOfSubject
}

func initialize() {
	wasmBytes, _ := ioutil.ReadFile("glue/smartcore_wasi_lib.wasm")
	engine := wasmer.NewEngine()
	store := wasmer.NewStore(engine)

	// Compiles the module
	module, _ := wasmer.NewModule(store, wasmBytes)

	wasiEnv, _ := wasmer.NewWasiStateBuilder("smartcore-wasi-go").
		PreopenDirectory(".").
		Finalize()

	importObject, _ := wasiEnv.GenerateImportObject(store, module)
	newInstance, _ := wasmer.NewInstance(module, importObject)

	Instance = newInstance
}

func getStringFromPointer(ptr int32, instance *wasmer.Instance) (string, int) {
	memory, _ := instance.Exports.GetMemory("memory")
	memoryData := memory.Data()[ptr:]
	len := len(memoryData)
	lengthOfString := 0
	for ; lengthOfString < len; lengthOfString++ {
		if memoryData[lengthOfString] == 0 {
			break
		}
	}
	return string(memoryData[0:lengthOfString]), lengthOfString
}

func Load_model() string {

	if Instance == nil {
		initialize()
	}
	load_model, _ := Instance.Exports.GetFunction("load_model")
	deallocate, _ := Instance.Exports.GetFunction("deallocate")
	modelPath, pathlen := getStringPointer("glue/iris_knn.model", Instance)
	outputPtr, _ := load_model(modelPath)
	output, outputLen := getStringFromPointer(outputPtr.(int32), Instance)
	deallocate(outputPtr, outputLen)
	deallocate(modelPath, pathlen)
	return output
}
